File: senior/MT1051/week06/HW16/HW16_109601003.py
# -*- coding: utf-8 -*-
"""
Date: 2023/10/23
HW: 16
Author: 林群賀
Student Number: 109601003
"""

def supply_price(P):
    return 43.3 + 1.25 * P

def demand_price(P):
    return 96.3 - 2.13 * P

def main() -> None:

    i = 0
    tolerance = 0.0001

    while True:
        Qs = supply_price(i)
        Qs = round(Qs, 3)
        Qd = demand_price(i)
        Qd = round(Qd, 3)

        if abs(Qs - Qd) < tolerance:
            print(f"均衡价格 (P) = {i:.3f}")
            print(f"均衡数量 (Q) = {Qs:.3f}")
            break

        i += 0.0001


if __name__ == "__main__":
    main()

File: other/MT1051/week06/HW16/HW16_109601003.go
/*
Created Date: 2023/10/23
HW: 16
Author: 林群賀
Student Number: 109601003
*/

package main

import (
	"fmt"
	"math"
)

func supplyPrice(P float64) float64 {
	return 43.3 + 1.25*P
}

func demandPrice(P float64) float64 {
	return 96.3 - 2.13*P
}

func main() {
	i := 0.0
	tolerance := 0.0001

	for {
		Qs := supplyPrice(i)
		Qs = round(Qs, 3)
		Qd := demandPrice(i)
		Qd = round(Qd, 3)

		if delta := Qs - Qd; delta < tolerance && delta > -tolerance {
			fmt.Printf("均衡价格 (P) = %.3f\n", i)
			fmt.Printf("均衡数量 (Q) = %.3f\n", Qs)
			break
		}

		i += 0.0001
	}
}

func round(x float64, decimals int) float64 {
	pow := math.Pow(10, float64(decimals))
	return math.Round(x*pow) / pow
}
